File: c/collatz-conjecture/collatz_conjecture.c
#include "collatz_conjecture.h"

int steps(int n)
{
    if (n < 1) {
        return ERROR_VALUE;
    } else if (n == 1) {
        return 0;
    }

    return 1 + steps(n % 2 == 0 ? n / 2 : 3 * n + 1);
}

File: csharp/collatz-conjecture/CollatzConjecture.cs
using System;
using System.Collections.Generic;
using System.Linq;

public static class CollatzConjecture
{
    public static int Steps(int number)
    {
        ArgumentOutOfRangeException.ThrowIfNegativeOrZero(number, nameof(number));

        return Collatz(number).Count();
    }

    private static IEnumerable<int> Collatz(int number)
    {
        while (number != 1)
        {
            yield return number;
            number = number % 2 == 0 ? number / 2 : number * 3 + 1;
        }
    }
}
